text/0249-v2-experiments.md

```diff
@@ -373,6 +367,363 @@ Disadvantages:
+```
+
+**If there are no breaking changes, are the APIs really experimental?**
+
+Given that preview APIs are safe to use, and no breaking changes will be
+introduced to them, we might consider not referring to them in any special way,
+and if needed, simply add a version to the API name. The first version of
+`grantWrite` will be named `grantWrite`, the second version will be named
+`grantWriteV1` and so on. While a preview API will not break, we should still
+use a naming scheme that convey its non-final nature for the following reasons:
+
+1. **Real-estate**: `grantWrite` is a much better name than `grantWritePre3` for
+   the final API.
+2. **Encourage feedback**: Declaring an API as "in preview" encourage the
+   community to supply feedback.
+3. **Setting the right expectations**: When an API in preview users are aware
+   that this is not the final version of the API, and its deprecation is
+   expected.
+
+### Modules lifecycle
+
+This section discuss adding a new modules to `aws-cdk-lib`.
+
+#### Alpha (optional)
+
+This stage is intended for modules that are in early development stage, they
+change rapidly and may incurs many breaking changes. These modules will be
+released separately from `aws-cdk-lib` and will have their own version, which
+allows for breaking changes in accordance to semver, e.g `0.x` or prerelease
+qualifiers.
+
+In order to release a module as alpha, the following conditions must be met:
+
+1. Module from `aws-cdk-lib` **can not** depend on it.
+2. Can not depend on any other alpha module\*\*.
+
+The purpose of the first condition is to prevent cyclic dependencies. To
+illustrate the purpose of the second condition lets look at an example. Assume
+we have an alpha module A, which depends on an alpha module B, graduating module
+A means it will be added to `aws-cdk-lib`, since module A depends on module B,
+it means that adding module A to `aws-cdk-lib` breaks condition 1 and creates a
+cyclic dependency. To prevent the cycle we will have to graduate B as well,
+which is not necessarily what we want. Additionally, if module B is an alpha
+module, module A will have to declare a fixed dependency on it, locking its
+consumers to the same version of module B - defeating the purpose of
+peerDependencies.
+
+> \*\* _There might be exceptions that will be considered on a per case basis_
+
+#### Beta (optional)
+
+In this stage modules are added to `aws-cdk-lib` with a preview suffix in the
+name of the module, e.g `aws-batch-beta-x`. Modules in this stage are not
+allowed to introduce any breaking changes to their API. Any non backward
+compatible change will be introduced via deprecation\*\*.
+
+> \*\* _The deprecation process will be discussed in the API previews
+> specification_
+
+#### Stable
+
+In this stage modules are added to `aws-cdk-lib` under their final name, e.g
+`aws-batch`. Addition of new APIs should follow the API previews specification.
```

Comment:
I don't see a difference between beta and stable. API breakages are not allowed in either case. can we just drop beta?

Reply:
Stable modules will have a stable name e.g  `aws-s3` , non final APIs within these modules will be add with the preview suffix, e.g `grantReadPre1`. Beta modules will have a suffix in their name, e.g `aws-s3-beta-1` but most** of their APIs will not have preview names.

The stable stage is intended for well baked modules with limited number of preview APIs. The beta stage is mainly intended for adding new modules. New modules are likely to change a lot, and as result will have many deprecated APIs.  Creating the module under a preview name allows us to deprecate the entire module when we are happy with the result, starting with a clean slate. This will effect the in IDE autocomplete experience, how the documentation looks and more. 

The ** explained:
We still need to finalize how will individual APIs be added inside the beta module. If we add them with a non preview name as suggested above, it will make the migration from beta to stable easy for customers - less code change. 

Adding an API to the  `aws-s3-beta-1` module:

```
grantWrite()
```

When we graduate the module, the name of the method will remain `grantWrite`, for modules with a large API surface area, this will have a substantial effect on the migration in terms of required code change. However, this assume only one breaking change for each API - the switch between beta to stable, which might not be true. It could be that during the beta phase of a module we will need to make a non backward compatible change to one of its API. Im not yet sure what's the right answer here, my current tendency is to assume that the common use case for beta development will be satisfied by a single non compatible change per API, and in cases where it is not, we can change the API name, e.g add `V2`. 

It will also make it easier to create a new L2, not having to add the preview suffix to all properties, class and methods.  But as I said, Im still thinking about it, feedback is welcome.

Reply:
@nija-at

Reply:
> The beta stage is mainly intended for adding new modules. New modules are likely to change a lot, and as result will have many deprecated APIs.

Wouldn't it be simpler if we just have all new modules start off as individual packages in the `aws-cdk-previews` (or whatever name we decide) repo?

It's not clear when one decides to make a beta package vs a package in `aws-cdk-previews` repo. Maybe I missed it?
It would be simpler to just have one.

Reply:
It will not be possible for all new modules, for example, modules with stable dependents must be added to `aws-cdk-lib`. 
The same is true for v1 experimental modules that must be add to `aws-cdk-lib` for the same reason, here is a list of all experimental modules that have stable dependent: 

```
@aws-cdk/aws-s3-assets		
@aws-cdk/aws-ecr-assets		
@aws-cdk/cx-api	
@aws-cdk/region-info		
@aws-cdk/aws-efs	
@aws-cdk/aws-batch	
@aws-cdk/aws-glue	
@aws-cdk/aws-ses	
@aws-cdk/aws-autoscaling-common
```

Previous analysis of these modules done by the construct library squad determined that graduating some of these modules will require a lot of work. The beta stage allows us to add these modules to `aws-cdk-lib` without having to graduate them.

Reply:
Let's re-visit these and see if we can make these stable. 
Some of these have been around for a lonnngg time (eg. s3-assets, ecr-assets, region-info, autoscaling-common) and I don't see why we shouldn't just call them 'stable'. Others like efs, glue and ses will need to be looked at.

Another option is to make these a one time 'beta' package, and never any other package. I'd rather not have yet another option available unless there's a really good motivation.

Reply:
Agreed. I think many of these can easily be stabilized without much additional effort given they've been baked for a very long time.

Reply:
I think these:

```
@aws-cdk/aws-s3-assets		
@aws-cdk/aws-ecr-assets		
@aws-cdk/cx-api	
@aws-cdk/region-info	
@aws-cdk/aws-autoscaling-common
```

are slam-dunks, and should be stabilized with v2.

These:

```
@aws-cdk/aws-efs	
@aws-cdk/aws-batch	
@aws-cdk/aws-glue	
@aws-cdk/aws-ses
```

are more tricky, and will probably require a separate decision for each one.

Reply:
There are two separate discussions here:

1. Do we need the beta phase in general.
2. How should v1 experimental modules should be introduce in v2.

Lets keep the discussion about 2 to our planned API review meeting. For 1, happy to hear feedback and refine the RFC

Reply:
I don't think we need a general beta phase. I've not heard a good argument for having such a mechanism for future modules.

As for the existing modules, if the dependent modules don't expose any of the interfaces from their experimental dependencies, we can safely remove the dependency and use the L1 (or escape hatch).
Another options to consider could be to deprecate these APIs (in favour of APIs that don't expose the underlying experimental module) and remove them for v2.

We can make a one-time exception for a couple of these packages if none of these options are simple.

Reply:
Good arguments for having a beta phase:

1. New modules that have stable dependencies and thus must be in `aws-cdk-lib`

Reply:
> Good arguments for having a beta phase:
> 
> 1. New modules that have stable dependencies and thus must be in `aws-cdk-lib`

Did you perhaps mean "dependents", not "dependencies"? Because stable dependencies are not a problem I don't think (all of the experimental modules will have a peer dependency on `aws-cdk-lib`, and so have access to all of the stable dependencies).

Reply:
I did - edited

Reply:
Did I miss something? I believe "beta" is equivalent to our current "developer preview" (or "release candidate"). My understanding is that modules start as "alpha" outside of aws-cdk-lib and when they are ready, they are added to the main library as "beta" to bake. Once finalized, the "beta" marker is removed and we call them stable. No?

Reply:
Per my discussion with @NetaNir on Friday, that's not the case. Modules can start either as alpha in the alpha repo or as beta in aws-cdk-lib. I'll let her expand on this.

@eladb - when they move to 'beta' within aws-cdk-lib, would the APIs be suffixed with the preview marker or stay the same? how would we change the APIs if we want to change something during the bake period? It's not clear to how we'll manage this 'beta' stage.

Reply:
beta1 beta2...